src/neorados/cls/fifo.h

```diff
@@ -0,0 +1,1514 @@
+	  nullptr,
+	  ca::bind_ea(
+	    e, a,
+	    [this, part_oid](bs::error_code ec) mutable {
+	      if (ec == bs::errc::no_such_file_or_directory) {
+		auto e = ba::get_associated_executor(handler,
+						     f->get_executor());
+		auto a = ba::get_associated_allocator(handler);
+		f->_read_meta(
+		  ca::bind_ea(
+		    e, a,
+		    [this](bs::error_code ec) mutable {
+		      if (ec) {
+			handle(ec);
+			return;
+		      }
+
+		      if (part_num < f->info.tail_part_num) {
+			/* raced with trim? restart */
+			max_entries += result.size();
+			result.clear();
+			part_num = f->info.tail_part_num;
+			ofs = 0;
+			list();
+		      }
+		      /* assuming part was not written yet, so end of data */
+		      more = false;
+		      handle({});
+		      return;
+		    }));
+		return;
+	      }
+	      if (ec) {
+		ldout(f->r->cct(), 0)
+		  << __func__
+		  << "(): list_part() on oid=" << part_oid
+		  << " returned ec=" << ec.message() << dendl;
+		handle(ec);
+		return;
+	      }
+	      if (ec_out) {
+		ldout(f->r->cct(), 0)
+		  << __func__
+		  << "(): list_part() on oid=" << f->info.part_oid(part_num)
+		  << " returned ec=" << ec_out.message() << dendl;
+		handle(ec_out);
+		return;
+	      }
+
+	      more = part_full || part_more;
+	      for (auto& entry : entries) {
+		list_entry e;
+		e.data = std::move(entry.data);
+		e.marker = marker{part_num, entry.ofs}.to_string();
+		e.mtime = entry.mtime;
+		result.push_back(std::move(e));
+	      }
+	      max_entries -= entries.size();
+	      entries.clear();
+	      if (max_entries > 0 &&
+		  part_more) {
+		list();
+		return;
+	      }
+
+	      if (!part_full) { /* head part is not full */
+		handle({});
+		return;
+	      }
+	      ++part_num;
+	      ofs = 0;
+	    list();
+	    }));
+      } else {
+	handle({});
+	return;
+      }
+    }
+  };
+
+  template<typename CT>
+  auto list(int max_entries,
+	    std::optional<std::string_view> markstr,
+	    CT&& ct) {
+    ba::async_completion<CT, void(bs::error_code,
+				  std::vector<list_entry>, bool)> init(ct);
+    std::unique_lock l(m);
+    std::int64_t part_num = info.tail_part_num;
+    l.unlock();
+    std::uint64_t ofs = 0;
+    auto a = ba::get_associated_allocator(init.completion_handler);
+    auto e = ba::get_associated_executor(init.completion_handler);
+
+    if (markstr) {
+      auto marker = to_marker(*markstr);
+      if (!marker) {
+	ldout(r->cct(), 0) << __func__
+			   << "(): failed to parse marker (" << *markstr
+			   << ")" << dendl;
+	e.dispatch(ca::bind_handler(std::move(init.completion_handler),
+				    errc::invalid_marker,
+				    std::vector<list_entry>{}, false), a);
+	return init.result.get();
+      }
+      part_num = marker->num;
+      ofs = marker->ofs;
+    }
+
+    using handler_type = decltype(init.completion_handler);
+    auto ls = ceph::allocate_unique<Lister<handler_type>>(
+      a, this, part_num, ofs, max_entries,
+      std::move(init.completion_handler));
+    ls.release()->list();
```

Comment:
doesn't this have the same issue as before, where we only free the memory if the completion handler gets called?

Reply:
Fixed.